### DISCUSSION TITLE
UI Widget configuration via msg.ui_control

### DISCUSSION THREAD

**Discussion by Sineos** [phoddie/node-red-mcu#96] [Ideas]
It would be great to modify UI aspect during run time. NR offers this with the `ui_control` properties: https://github.com/node-red/node-red-dashboard/blob/master/config-fields.md

I tried this for the gauge-node and it seems not implemented so far.

Note:
Should this be implemented the gauge-node would need some modification as well. If just a  `msg.ui_control` is generated without any `msg.payload` then NR sets `msg.payload = undefined`. In the gauge-node this leads to:
`C:\Users\souls\Projects\moddable\modules\commodetto\outline\commodettoOutline.c (157) # Break: arc: invalid endAngle!`

**Comment by phoddie** [comment]
Ah, the many lurking features of Node-RED. ;) The single `ui_control` property expands out to a vast set of features. No doubt some are straightforward to support, others impractical, and many in between. Perhaps a good starting point is to try to identify a few that would be valuable to have and then evaluate the implementation complexity.

**Reply by Sineos** [reply]
Of course reflecting my personal preferences but also coming from the idea, what could be useful in an MCU project, I marked the properties in **bold**

|widget                 |property           |type       | notes / example
|---                    |---                |---        |---
|ui_button              |**color**              |string     | not needed
|                       |**bgcolor**         |string     | not needed
|                       |icon               |string     | on refresh
|                       |format             |string     | not needed
|                       |tooltip            |string     | on refresh
|                       |className          |string     | (Note 4)
|ui_chart               |look               |string     |"line","bar","horizontalBar","pie","polar-area","radar"
|                       |legend             |boolean    |&nbsp;
|                       |interpolate        |string     |"linear","step","bezier"
|                       |nodata             |string     |&nbsp;
|                       |**ymin**               |number     |&nbsp;
|                       |**ymax**               |number     |&nbsp;
|                       |dot                |boolean    |&nbsp;
|                       |xformat            |string     |"HH:mm:ss"
|                       |cutout             |number     |&nbsp;
|                       |**colors**             |object     | n/a
|                       |useOneColor        |boolean    | n/a
|                       |spanGaps           |boolean    | n/a
|                       |animation          |string     | (Note 1), {duration:1000, easing:"easeInOutSine"}
|                       |options            |object     | (Note 2), {scales: {yAxes: [{ticks: {fontSize: 20}}]}}
|                       |className          |string     | (Note 4)
|ui_colour_picker       |format             |string     | on refresh
|                       |showPicker         |boolean    | on refresh
|                       |showSwatch         |boolean    | on refresh
|                       |showValue          |boolean    | on refresh
|                       |showAlpha          |boolean    | on refresh
|                       |showLightness      |boolean    | on refresh
|                       |className          |string     | (Note 4)
|ui_dropdown            |place              |string     |"placeholder text"
|                       |**options**            |array      |[{"label":"foo","value":"0","type":"str"}]
|                       |className          |string     | (Note 4)
|ui_gauge               |gtype              |string     |"gage", "donut", "compass", "wave"
|                       |**min**                |number     |&nbsp;
|                       |**seg1**               |number     |segment 1 limit
|                       |**seg2**               |number     |segment 2 limit
|                       |**max**                |number     |&nbsp;
|                       |**colors**             |array      |["blue","#00ff00","#f00"]
|                       |options            |object     |(see Note 3 below)
|                       |className          |string     | (Note 4)
|ui_numeric             |**min**                |number     |&nbsp;
|                       |**max**                |number     |&nbsp;
|                       |step               |number     |&nbsp;
|                       |format             |string     |"{{value}}"
|                       |className          |string     | (Note 4)
|ui_slider              |min                |number     |&nbsp;
|                       |max                |number     |&nbsp;
|                       |step               |number     |&nbsp;
|                       |className          |string     | (Note 4)
|ui_switch              |onicon             |string     | all or nothing
|                       |officon            |string     | all or nothing
|                       |oncolor            |string     | all or nothing
|                       |offcolor           |string     | all or nothing
|                       |className          |string     | (Note 4)
|ui_template            |format             |string     | the script
|                       |templateScope      |string     | "local" or "global"
|                       |className          |string     | (Note 4)
|ui_text                |format             |string     |"{{value}}"
|                       |layout             |string     | "row-left", "row-right", etc
|                       |className          |string     | (Note 4)
|ui_text_input          |mode               |string     | "text", "email", "password", "color"
|                       |delay              |number     |&nbsp;
|                       |className          |string     | (Note 4)
